### PR TITLE
Make it validate input continuously

### PR DESCRIPTION
**Pull request by fadeevab**

### CODE REVIEW

**Comment by fadeevab** [comment]
@samuela Continuous validation works just fine, I don't see huge objections why not to enable it mandatory. It looks cool actually.

File: src/input.rs [hunk 1]
```diff
@@ -110,27 +110,24 @@ where
...
-            match self.input.to_string().parse::<T>() {
-                Ok(value) => return State::Submit(value),
-                Err(_) => {
-                    return State::Error("Invalid value format".to_string());
-                }
-            }
+        match self.input.to_string().parse::<T>() {
+            Ok(value) if *key == Key::Enter => return State::Submit(value),
+            Err(_) => return State::Error("Invalid value format".to_string()),
+            _ => {}
```
[the remark this answers is not in the document]
Comment (fadeevab) [reply src/input.rs:130]: Yes, per every key which is not Enter and not error...

**Comment by fadeevab** [comment]
@samuela Validation per every key may be lagging with more text input and with more complex validation. Some config would be nice to have probably. Let's delay this patch. I will check the spinner animation feature request, and if it's simple I will publish the update.

File: src/input.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (samuela) [reply src/input.rs:130]: oh oops, missed the if condition on the `Ok` branch

**Comment by samuela** [comment]
makes sense, sgtm!

**Comment by fadeevab** [comment]
@samuela I decided to add `validate_continuously`, I can go with this

**Comment by samuela** [comment]
Ah this is slightly different than the design I had in mind. I was thinking of something along the lines of
```rs
username_input
  .validate_interactive(|x| if x.len() > 3 { Ok(()) } else { Err("input not long enough") })
  .validate_on_enter(check_username_is_available)
```
That way users could specify any combination: no validation, interactive only, on-enter only, or interactive+on-enter validation.

That said, I'm not tied to any specific design and ultimately it's your choice of course :)

**Comment by fadeevab** [comment]
@samuela I had the same idea, but what's about `validate`? We have to keep it for backward compatibility. As an option, it could be the same as `validate_on_enter`. And seems like `validate_interactive` should NOT validate on Enter. Finally, if you wanna do a similar validation interactively and on_enter, then the same lambda should be re-used twice. Sounds right?

**Comment by fadeevab** [comment]
@samuela I did it, check it out :)

**Comment by samuela** [comment]
> Finally, if you wanna do a similar validation interactively and on_enter, then the same lambda should be re-used twice. Sounds right?

I think in this case `Submit` would not be possible as long as `validate_interactive` returns an `Err`. So it suffices to only set `validate_interactive`.

File: src/input.rs [hunk 2]
```diff
@@ -72,13 +73,35 @@ impl Input {
         self
     }
 
-    /// Sets a validation callback for the input.
+    /// Sets a validation callback for the input that is called when the user submits.
+    /// The same as [`Input::validate_on_enter`].
```
Comment (samuela) [inline src/input.rs:77]: Shall we mark as `#[deprecated]`?

File: src/input.rs [hunk 3]
```diff
@@ -72,13 +73,35 @@ impl Input {
...
     pub fn validate<V>(mut self, validator: V) -> Self
     where
         V: Validate<String> + 'static,
         V::Err: ToString,
     {
-        self.validate = Some(Box::new(move |input: &String| {
+        self.validate_on_enter = Some(Box::new(move |input: &String| {
```
Comment (samuela) [inline src/input.rs:83]: Given the naming convention here, I was expecting `validate()` to be deprecated and be a shim that calls out to `validate_on_enter()` as opposed to the other way around. Doesn't really matter functionality-wise, but just thought I'd raise the point.

**Review by samuela — APPROVED** [review]
Nice!

File: src/input.rs (hunk 3 above)
Comment (fadeevab) [reply src/input.rs:83]: I mean, the original idea was to keep `cliclack` close to `@clack/prompts`, so let it be without deprecation, it doesn't make any harm anyways.

**Comment by fadeevab** [comment]
> I think in this case Submit would not be possible as long as validate_interactive returns an Err. So it suffices to only set validate_interactive.

> EDIT: bright minds think alike -- seems like you came to same conclusion!

@samuela haha, yes, at some point it's arranged somehow naturally in the code :)

**Pull request merged**